### ISSUE TITLE
Inconsistency between local test and CI

### ISSUE THREAD

**Issue by blasrodri** [near/nearcore#7238]
**Describe the bug**
When running on commit 16a332d4fd908e83bbf601a074974ff98ffcf474: 
```sh
cargo test --locked -p 'integration-tests'
```
On CI the output is: https://buildkite.com/nearprotocol/nearcore/builds/18159#01822680-14d8-46e4-85e4-cdbaebef92ef
```
---- tests::runtime::state_viewer::test_view_state stdout ----
--
  | thread 'tests::runtime::state_viewer::test_view_state' panicked at 'assertion failed: `(left == right)`
  | left: `Some([Extension(TrieProofExtension { key: [16], child_hash: `u69gJXVUpfvGomSGzwbC9i5NU3VaSjPNqfREvbXysii`, memory_usage: 97478 }), Branch(TrieProofBranch { children: [Some(`Er6tjbuytQr2QvL5TjqkEdoC7ndDHPCcZUPTza9xvBcz`), Some(`E6YrvRisYFW9qZL4H7pThiRuYEjZ4mfB75vJArgMuTdg`), Some(`CHDm3ySyAKvcb6nctEaGfqjY7EGLz9tXJZ28XEXGDi6N`), None, None, None, None, None, None, Some(`FLxcnKrm7fPnZNdeVvybWtnQGBmyoUuXjhXmhq3NgNPa`), None, None, None, None, None, None], value: None, memory_usage: 97426, index: 9 }), Branch(TrieProofBranch { children: [Some(`Er6tjbuytQr2QvL5TjqkEdoC7ndDHPCcZUPTza9xvBcz`), Some(`E6YrvRisYFW9qZL4H7pThiRuYEjZ4mfB75vJArgMuTdg`), Some(`CHDm3ySyAKvcb6nctEaGfqjY7EGLz9tXJZ28XEXGDi6N`), None, None, None, None, None, None, None, None, None, None, None, None, None], value: None, memory_usage: 97426, index: 9 }), Extension(TrieProofExtension { key: [22, 22, 198], child_hash: `HCEH9UoyeZUv5UDgu9QpFTxWoFGPMrS4VAV2TZHzSLjw`, memory_usage: 780 }), Branch(TrieProofBranch { children: [None, None, None, None, None, Some(`GxqgVRkFxzLvar8iHeBADqu2PpRiH3fYVdPGzw7zUdzR`), None, None, None, Some(`HQRiU7i61GFKWtorr8RzY1QyQVJqX5hrytYqwhiyHZHd`), None, None, None, None, None, None], value: None, memory_usage: 724, index: 9 }), Branch(TrieProofBranch { children: [None, None, None, None, None, Some(`GxqgVRkFxzLvar8iHeBADqu2PpRiH3fYVdPGzw7zUdzR`), None, None, None, None, None, None, None, None, None, None], value: None, memory_usage: 724, index: 9 }), Extension(TrieProofExtension { key: [22], child_hash: `GqYVVbkA7nvAEuCsvKZdWQ7Ddc5e9bNsfVhTTusWyvup`, memory_usage: 559 }), Branch(TrieProofBranch { children: [None, None, None, Some(`5H3fJHA6jCZKnbFnkFDcuPDRbPdmiSgC4mR4jX4D5XD3`), None, None, None, None, None, None, None, None, None, None, Some(`EuTKawKwL1A9pgVPxV1EHpzFYYBxjy9DP55VkcWX4GQP`), None], value: None, memory_usage: 507, index: 3 }), Branch(TrieProofBranch { children: [None, None, None, None, None, None, None, None, None, None, None, None, None, None, Some(`EuTKawKwL1A9pgVPxV1EHpzFYYBxjy9DP55VkcWX4GQP`), None], value: None, memory_usage: 507, index: 3 }), Extension(TrieProofExtension { key: [22, 82, 230, 230, 86, 23, 34, 199, 70, 87, 55, 67], child_hash: `EeJvs6otTUKtsbpfcgm59ePFGw7QMB5G37hAVAS6n341`, memory_usage: 342 }), Branch(TrieProofBranch { children: [None, Some(`6kH6uEShmyoX7iGisy8nZxoYVNw6w8McqqDgXHct7J6X`), None, Some(`7H2oXQXYxnuB3Kq6djTnS7rkTdsgmLhByL3JFHLsrFQM`), None, None, None, None, None, None, None, None, None, None, None, None], value: None, memory_usage: 268, index: 1 }), Branch(TrieProofBranch { children: [None, None, None, Some(`7H2oXQXYxnuB3Kq6djTnS7rkTdsgmLhByL3JFHLsrFQM`), None, None, None, None, None, None, None, None, None, None, None, None], value: None, memory_usage: 268, index: 1 }), Leaf(TrieProofLeaf { key: [32, 50, 51], value_length: 3, value_hash: `CCYa6DM7NuL4iPoSM4PvtYMRcHFydKhD5oRSXx3jfxCS`, memory_usage: 109 })])`,
  | right: `Some([Extension(TrieProofExtension { key: [16], child_hash: `FwUXNsdcWaTQdZHgmHDjXuCGPmS8f4veZEKSpnNwFL6e`, memory_usage: 97430 }), Branch(TrieProofBranch { children: [Some(`8A6QyKBCCAAtVAeSJcgGtUaFQZCdpd4W8hG9kHNV67WH`), Some(`Ci8TDPmfawsHAVC9tUCvRuJKUTAhPJGrpgGbxKNBCroY`), Some(`CHDm3ySyAKvcb6nctEaGfqjY7EGLz9tXJZ28XEXGDi6N`), None, None, None, None, None, None, Some(`FLxcnKrm7fPnZNdeVvybWtnQGBmyoUuXjhXmhq3NgNPa`), None, None, None, None, None, None], value: None, memory_usage: 97378, index: 9 }), Branch(TrieProofBranch { children: [Some(`8A6QyKBCCAAtVAeSJcgGtUaFQZCdpd4W8hG9kHNV67WH`), Some(`Ci8TDPmfawsHAVC9tUCvRuJKUTAhPJGrpgGbxKNBCroY`), Some(`CHDm3ySyAKvcb6nctEaGfqjY7EGLz9tXJZ28XEXGDi6N`), None, None, None, None, None, None, None, None, None, None, None, None, None], value: None, memory_usage: 97378, index: 9 }), Extension(TrieProofExtension { key: [22, 22, 198], child_hash: `HCEH9UoyeZUv5UDgu9QpFTxWoFGPMrS4VAV2TZHzSLjw`, memory_usage: 780 }), Branch(TrieProofBranch { children: [None, None, None, None, None, Some(`GxqgVRkFxzLvar8iHeBADqu2PpRiH3fYVdPGzw7zUdzR`), None, None, None, Some(`HQRiU7i61GFKWtorr8RzY1QyQVJqX5hrytYqwhiyHZHd`), None, None, None, None, None, None], value: None, memory_usage: 724, index: 9 }), Branch(TrieProofBranch { children: [None, None, None, None, None, Some(`GxqgVRkFxzLvar8iHeBADqu2PpRiH3fYVdPGzw7zUdzR`), None, None, None, None, None, None, None, None, None, None], value: None, memory_usage: 724, index: 9 }), Extension(TrieProofExtension { key: [22], child_hash: `GqYVVbkA7nvAEuCsvKZdWQ7Ddc5e9bNsfVhTTusWyvup`, memory_usage: 559 }), Branch(TrieProofBranch { children: [None, None, None, Some(`5H3fJHA6jCZKnbFnkFDcuPDRbPdmiSgC4mR4jX4D5XD3`), None, None, None, None, None, None, None, None, None, None, Some(`EuTKawKwL1A9pgVPxV1EHpzFYYBxjy9DP55VkcWX4GQP`), None], value: None, memory_usage: 507, index: 3 }), Branch(TrieProofBranch { children: [None, None, None, None, None, None, None, None, None, None, None, None, None, None, Some(`EuTKawKwL1A9pgVPxV1EHpzFYYBxjy9DP55VkcWX4GQP`), None], value: None, memory_usage: 507, index: 3 }), Extension(TrieProofExtension { key: [22, 82, 230, 230, 86, 23, 34, 199, 70, 87, 55, 67], child_hash: `EeJvs6otTUKtsbpfcgm59ePFGw7QMB5G37hAVAS6n341`, memory_usage: 342 }), Branch(TrieProofBranch { children: [None, Some(`6kH6uEShmyoX7iGisy8nZxoYVNw6w8McqqDgXHct7J6X`), None, Some(`7H2oXQXYxnuB3Kq6djTnS7rkTdsgmLhByL3JFHLsrFQM`), None, None, None, None, None, None, None, None, None, None, None, None], value: None, memory_usage: 268, index: 1 }), Branch(TrieProofBranch { children: [None, None, None, Some(`7H2oXQXYxnuB3Kq6djTnS7rkTdsgmLhByL3JFHLsrFQM`), None, None, None, None, None, None, None, None, None, None, None, None], value: None, memory_usage: 268, index: 1 }), Leaf(TrieProofLeaf { key: [32, 50, 51], value_length: 3, value_hash: `CCYa6DM7NuL4iPoSM4PvtYMRcHFydKhD5oRSXx3jfxCS`, memory_usage: 109 })])`', integration-tests/src/tests/runtime/state_viewer.rs:153:5
  | stack backtrace:
  | 0: rust_begin_unwind
  | at /rustc/a8314ef7d0ec7b75c336af2c9857bfaf43002bfc/library/std/src/panicking.rs:584:5
  | 1: core::panicking::panic_fmt
  | at /rustc/a8314ef7d0ec7b75c336af2c9857bfaf43002bfc/library/core/src/panicking.rs:142:14
  | 2: core::panicking::assert_failed_inner
  | 3: core::panicking::assert_failed
  | at /rustc/a8314ef7d0ec7b75c336af2c9857bfaf43002bfc/library/core/src/panicking.rs:181:5
  | 4: integration_tests::tests::runtime::state_viewer::test_view_state
  | at ./src/tests/runtime/state_viewer.rs:153:5
  | 5: integration_tests::tests::runtime::state_viewer::test_view_state::{{closure}}
  | at ./src/tests/runtime/state_viewer.rs:110:1
  | 6: core::ops::function::FnOnce::call_once
  | at /rustc/a8314ef7d0ec7b75c336af2c9857bfaf43002bfc/library/core/src/ops/function.rs:248:5
  | 7: core::ops::function::FnOnce::call_once
  | at /rustc/a8314ef7d0ec7b75c336af2c9857bfaf43002bfc/library/core/src/ops/function.rs:248:5
  | note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
  |  
  |  
  | failures:
  | tests::runtime::state_viewer::test_view_state


```
On my machine `AMD Ryzen 9 5900X 12-Core Processor`, and with `rustc 1.62.0 (a8314ef7d 2022-06-27)`

```
test result: ok. 207 passed; 0 failed; 77 ignored; 0 measured; 0 filtered out; finished in 65.12s

   Doc-tests integration-tests

running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s
```

**Expected behavior**
The same output as CI. I am unsure of what the source of non-determinism can be.

**Comment by akhi3030** [comment]
If you try restarting the CI job, does it always fail?  i.e. does the CI fail consistently? And similarly does the test pass consistently on your machine?

**Comment by akhi3030** [comment]
Are you using the same version of the toolchain as CI?  I don't expect that to make a difference but if that is making a difference, then that is a venue we should explore.

**Comment by akhi3030** [comment]
Related PR: https://github.com/near/nearcore/pull/7205

**Comment by akhi3030** [comment]
So the line that the test is failing at is part of the PR: https://github.com/near/nearcore/pull/7205/files#diff-f5f96a83dcdc3bbdbd9eeefbf6a60b1f9a623b4b3231b9429dda5e0060233f1cR153.  I see that you hard coded some hash values in that line.  How did you compute the hash values?  Maybe the problem is that the way you computed them is different than how the CI is computing them.

**Comment by blasrodri** [comment]
> Are you using the same version of the toolchain as CI? I don't expect that to make a difference but if that is making a difference, then that is a venue we should explore.

Isn't it enforced here? https://github.com/near/nearcore/blob/master/rust-toolchain.toml

**Comment by akhi3030** [comment]
> > Are you using the same version of the toolchain as CI? I don't expect that to make a difference but if that is making a difference, then that is a venue we should explore.
> 
> Isn't it enforced here? https://github.com/near/nearcore/blob/master/rust-toolchain.toml

Of course, then this shouldn't be the problem.

**Comment by blasrodri** [comment]
> So the line that the test is failing at is part of the PR: https://github.com/near/nearcore/pull/7205/files#diff-f5f96a83dcdc3bbdbd9eeefbf6a60b1f9a623b4b3231b9429dda5e0060233f1cR153. I see that you hard coded some hash values in that line. How did you compute the hash values? Maybe the problem is that the way you computed them is different than how the CI is computing them.

I first checked the result I was getting, and copied it. I'm not 100% of what the real value should be. Regardless of that, the value should be the same in the different environments as far as I expect.

**Comment by blasrodri** [comment]
A recording of my local setup.
Passes on rust stable, and fails on nightly:
https://asciinema.org/a/pW3IVgfEuk8Ukp0WFS8b0f3WS @akhi3030

**Comment by matklad** [comment]
Hm, for me locally `cargo t -p integration-tests -- test_view_state` fails consistently, so it's not specific to CI

**Comment by blasrodri** [comment]
> Hm, for me locally `cargo t -p integration-tests -- test_view_state` fails consistently, so it's not specific to CI

Can you check the cinema link?

**Comment by matklad** [comment]
Debugged this. This line is non-deterministic: 

https://github.com/matklad/nearcore/blob/16a332d4fd908e83bbf601a074974ff98ffcf474/integration-tests/src/runtime_utils.rs#L23

Here, we are adding a freshly-compiled contract to the state, and, as we compile the contract from Rust, and Rust builds are not fully hermetic yet, this depends on versions of compilers, your user name and whatnot. 

For the tests, we should ensure that, if we deploy a contract, it's a deterministic wasm blob. Cursory look shows that we might not need to deploy a contract for this test at all perhaps?

**Comment by blasrodri** [comment]
@matklad good catch. Fixed in [c7c1e0f](https://github.com/near/nearcore/pull/7205/commits/c7c1e0f2ab12afe9434f0aa0b258b100da75e8c4) in https://github.com/near/nearcore/pull/7205
